### PR TITLE
Add triage, which clusters together failures to be kinder to humans.

### PR DESCRIPTION
**Pull request by rmmh**
This is the first part, which loads failures from BigQuery, digests it
into manageable clusters, and uploads the resulting JSON to a GCS bucket.

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -6,3 +6,21 @@ git_repository(
...
+load("@org_pubref_rules_node//node:rules.bzl", "node_repositories", "npm_repository")
+
+node_repositories()
+
+npm_repository(
+    name = "npm_mocha",
+    deps = {
+        "mocha": "3.2.0",
+    },
+    # sha256 = "e43184e92109a2d4108f14cbd9316e2af05dd728697e3909e2d6a9f61d7ce395",
```
Comment (spxtr) [inline WORKSPACE]: Why commented out?

File: triage/update_summaries.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+set -exu
+cd $(dirname $0)
+table_mtime=$(bq --format=json show 'k8s-gubernator:build.week' | jq -r '(.lastModifiedTime|tonumber)/1000|floor' )
+if [[ ! -e triage_builds.json ]] || [ $(stat -c%Y triage_builds.json) -lt $table_mtime ]; then
+  echo "UPDATING" $table_mtime
+  bq --headless --format=json query -n 1000000 'select path, timestamp_to_sec(started) started, elapsed, tests_run, tests_failed, result, executor, job, number from [k8s-gubernator:build.week]' > triage_builds.json
+  bq --headless --format=json query -n 10000000 'select path build, test.name name, test.failure_text failure_text from [k8s-gubernator:build.week] where test.failed' > triage_tests.json
+  rm -f failed*.json
+fi
+pypy summarize.py triage_builds.json triage_tests.json
```
Comment (spxtr) [inline triage/update_summaries.sh]: Out of interest, how big of a difference does pypy make?

**Comment by spxtr** [comment]
Can you add more comments to the javascript? It's super hard to follow.

**Comment by fejta** [comment]
Very cool! Some style/readability suggestions mostly

**Comment by fejta** [comment]
FYI the changes so far are fantastic!

 https://storage.googleapis.com/k8s-gubernator/triage/index.html#72f3709fce203fbeef84 dropped from 300/hr to 13/hr. Can we a) Display 300/hr instead of 47443 weekly failures in the top and b) sort these by the hourly rate for the past day? Despite falling from 300 to 13 this one still shows up at the top. If I fix it today I'd like to see it move down tomorrow (maybe show 13/hr today, 300/hr this week or whatever)

**Comment by fejta** [comment]
Does this have analytics in it yet?

**Comment by rmmh** [comment]
Now it does!

File: triage/script.js [hunk 3]
```diff
@@ -0,0 +1,738 @@
...
+var lastClusterRendered = 0;  // for infinite scrolling
+
+const kGraphHeight = 200; 	  // keep synchronized with style.css:div.graph
+
+/**********************************************************************
+ *                        POLYFILLS                                   *
+ **********************************************************************/
+
+if (Object.entries === undefined) {
+	// Simple polyfill for Safari compatibility.
```
Comment (fejta) [inline triage/script.js]: What is polyfill?

File: triage/script.js [hunk 4]
```diff
@@ -0,0 +1,738 @@
...
+
+// Store information about individual builds.
+class Builds {
+	constructor(dict) {
+		this.jobs = dict.jobs;
+		this.jobPaths = dict.job_paths;
+		this.cols = dict.cols;
+		this.colStarted = this.cols.started;
+		this.colPr = this.cols.pr;
+		this.timespan = minMaxArray(this.cols.started);
```
Comment (fejta) [inline triage/script.js]: Would you want this to be started + duration?

File: triage/script.js [hunk 5]
```diff
@@ -0,0 +1,738 @@
...
+		this.runCount = this.cols.started.length;
+	}
+
+	// Create a build object given a job and build number.
+	get(job, number) {
+		let indices = this.jobs[job];
+		if (indices.constructor === Array) {
```
Comment (fejta) [inline triage/script.js]: Why this.jobs[job] sometimes an array and sometimes a number?

File: triage/script.js [hunk 6]
```diff
@@ -0,0 +1,738 @@
...
+		}
+		// Add more columns as necessary.
+		// This is faster than dynamically adding properties to an object.
+		return {
+			job: job,
+			number: number,
+			started: this.colStarted[index],
+			pr: this.colPr[index],
+		};
+		/* Old way:
```
Comment (fejta) [inline triage/script.js]: ?

File: triage/script.js [hunk 7]
```diff
@@ -0,0 +1,738 @@
...
+	}
+}
+
+function sum(arr, keyFunc) {
+	if (arr.length === 0)
+		return 0;
+	return arr.map(keyFunc).reduce((a, b) => a + b);
+}
+
+var clustersSum = clusters => sum(clusters, c => sum(c[1], t => t[1].length));
```
Comment (fejta) [inline triage/script.js]: Will you please

a) limit anonymous function depth to 1 (aka do not have an anon function inside an anon function) and
b) consider using `{ }` and indent by two spaces? (like what you're doing with `vals.sort` below

Also any reason not to make this a normal function?

File: triage/script.js [hunk 8]
```diff
@@ -0,0 +1,738 @@
...
+		this.sum = sum(this.c, c => clustersSum(c[3]));
+		this.cById = {};
+		for (let cluster of this.c) {
+			let keyId = cluster[1];
+			if (!this.cById[keyId]) {
+				this.cById[keyId] = cluster;
+			}
+		}
+		this.sumRecent = 0;
+		if (sortBy !== undefined) {
```
Comment (fejta) [inline triage/script.js]: When is sortBy undefined?

File: triage/script.js [hunk 9]
```diff
@@ -0,0 +1,738 @@
...
+			var keyFunc = {
+				total: c => [-clustersSum(c[3])],
+				message: c => [c[0]],
+				day: c => [-this.getHitsInLastDay(c[1]), -clustersSum(c[3])],
```
Comment (fejta) [inline triage/script.js]: Do we calculate clustersSum twice? Should we?

File: triage/script.js [hunk 10]
```diff
@@ -0,0 +1,738 @@
...
+			if (elA < elB) return -1;
+		}
+	});
+	return vals.map(x => x[1]);
+}
+
+// Store test clusters and support iterating and refiltering through them.
+class Clusters {
+	constructor(clustered, sortBy) {
+		this.c = clustered;
```
Comment (fejta) [inline triage/script.js]: Can we use an easier to understand name than c?

File: triage/script.js [hunk 11]
```diff
@@ -0,0 +1,738 @@
...
+				clustersOut = sortByKey(clustersOut, c => [-sum(c[1], t => t[1].length)]);
+				out.push([key, keyId, text, clustersOut]);
+			}
+		}
+		return new Clusters(out, opts.sort);
+	}
+}
+
+/**********************************************************************
+ *                           RENDERING                                *
```
Comment (fejta) [inline triage/script.js]: Any reason not to move this into its own file?

File: triage/script.js [hunk 12]
```diff
@@ -0,0 +1,738 @@
...
+// One-time initialization of the whole page.
+function load() {
+	setOptionsFromURL();
+	google.charts.load('current', {'packages': ['corechart', 'line']});
+	google.charts.setOnLoadCallback(() => { google.charts.loaded = true });
+
+	var setLoading = t => document.getElementById("loading-progress").innerText = t;
+	var toMB = b => Math.round(b / 1024 / 1024 * 100) / 100;
+
+	get('/k8s-gubernator/triage/failure_data.json',
```
Comment (fejta) [inline triage/script.js]: IMO this would be easier to read if it followed python style (indent the args from `get()`)

File: triage/script.js [hunk 13]
```diff
@@ -0,0 +1,738 @@
...
+}
+
+// Download a file from GCS and invoke callback with the result.
+// extracted/modified from kubernetes/test-infra/gubernator/static/build.js
+function get(uri, callback, onprogress) {
+	if (uri[0] === '/') {
+		// Matches /bucket/file/path -> [..., "bucket", "file/path"]
+		var groups = uri.match(/([^/:]+)\/(.*)/);
+		var bucket = groups[1], path = groups[2];
+		var url = 'https://www.googleapis.com/storage/v1/b/' + bucket + '/o/' +
```
Comment (fejta) [inline triage/script.js]: Why not use the xml api for this? (aka storage.googleapis.com/<bucket>/<path>)

File: triage/script.js [hunk 14]
```diff
@@ -0,0 +1,738 @@
...
+		setTimeout(() => {
+			builds = new Builds(data.builds);
+			clusteredAll = new Clusters(data.clustered);
+			rerender();
+			if (window.location.hash) {
+				renderUntilFound(window.location.hash.slice(1));
+			}
+		}, 0);
+	},
+	evt => {
```
Comment (fejta) [inline triage/script.js]: Very cool!

File: triage/script.js [hunk 15]
```diff
@@ -0,0 +1,738 @@
...
+		if (shown >= count) break;
+	}
+}
+
+// Clear the page and reinitialize the renderer and filtering. Render a few failures.
+function rerender(maxCount) {
+	if (!clusteredAll) return;
+
+	options = readOptions();
+	clustered = clusteredAll.refilter(options);
```
Comment (fejta) [inline triage/script.js]: Should we refilter if the options are the same?

File: triage/script.js [hunk 16]
```diff
@@ -0,0 +1,738 @@
...
+			text.textContent = text.textContent.replace(downArrow, rightArrow);
+			child.style = "display: none";
+			return true;
+		}
+	}
+	return false;
+}
+
+/**********************************************************************
+ *                           USER INPUT                               *
```
Comment (fejta) [inline triage/script.js]: Separate script?

File: triage/berghelroach.py [hunk 17]
```diff
@@ -0,0 +1,388 @@
...
+# </pre>
+# (We also allow f(p+2,p)=f(-p-2,p)=-1, causing those values to
+# have no effect in the starting row computation.]
+#
+# We only expand the set of diagonals visited every other round,
+# when (p-main) or (p+main) is even.  We keep track of even/oddness
+# to save some arithmetic.  The first round is always even, as p=abs(main).
+# Note that we rename the "f" function to "computeRow" to be Googley.
+
+class BerghelRoach(object):
```
Comment (fejta) [inline triage/berghelroach.py]: unit tests?

File: triage/script.js (hunk 3 above)
Comment (rmmh) [inline triage/script.js]: "In web development, a polyfill is code that implements a feature on web browsers that do not support the feature. "

File: triage/script.js (hunk 13 above)
Comment (rmmh) [inline triage/script.js]: The XML API doesn't serve up Access-Control-Allow-Origin headers to allow Javascript to make cross-origin requests.

File: triage/summarize.py [hunk 18]
```diff
@@ -0,0 +1,369 @@
...
+    return v1[len(t)]
+
+
+def make_ngram_counts(s, ngram_counts={}):
+    '''
+    Convert a string into a histogram of frequencies for different byte combinations.
+    This can be used as a heuristic to estimate edit distance between two strings in
+    constant time.
+    '''
+    size = 1 << 6
```
Comment (rmmh) [inline triage/summarize.py]: What do you mean ceiling?

File: triage/script.js (hunk 13 above)
Comment (fejta) [inline triage/script.js]: https://cloud.google.com/storage/docs/cross-origin

**Comment by fejta** [comment]
Please resolve my comments somehow to let me know when this is ready for another review!

File: triage/script.js (hunk 15 above)
Comment (rmmh) [inline triage/script.js]: rerender() is called on first load and on form changes, so options would necessarily be different in this case.

File: triage/script.js (hunk 9 above)
Comment (rmmh) [inline triage/script.js]: No, these are different sorting keys that can be used.

File: triage/script.js (hunk 13 above)
Comment (rmmh) [inline triage/script.js]: The JSON API is more appropriate for JS anyways, so why the need to switch?

File: triage/berghelroach.py (hunk 17 above)
Comment (rmmh) [inline triage/berghelroach.py]: done (ported from Java)

File: triage/script.js (hunk 11 above)
Comment (rmmh) [inline triage/script.js]: Moved.

File: triage/script.js (hunk 16 above)
Comment (rmmh) [inline triage/script.js]: Moved.

File: triage/script.js (hunk 13 above)
Comment (fejta) [inline triage/script.js]: Fair point. I just consider www.googleapis.com/storage/v1/b/foo/o/whatever considerably more gross than storage.googleapis.com/foo/whatever

Comment (fejta) [inline triage/script.js]: And the reason I suggest this is because you're just requesting the content, which has nothing to do with json

File: triage/script.js (hunk 10 above)
Comment (rmmh) [inline triage/script.js]: Done.

File: triage/script.js (hunk 7 above)
Comment (rmmh) [inline triage/script.js]: Done.

File: triage/script.js (hunk 6 above)
Comment (rmmh) [inline triage/script.js]: Removed.

File: triage/script.js (hunk 4 above)
Comment (rmmh) [inline triage/script.js]: No, since the start and end times are more useful for graphing and other information display.

File: triage/script.js (hunk 5 above)
Comment (rmmh) [inline triage/script.js]: It's coupled with builds_to_columns in Python: "optimization: if we have a dense sequential mapping of builds=>indexes, store only the first build number, the run length, and the first index number."

File: triage/script.js (hunk 8 above)
Comment (rmmh) [inline triage/script.js]: Refactored to do the sorting inside refilter.

**Comment by rmmh** [comment]
@fejta ready for a final round

File: triage/berghelroach.py [hunk 19]
```diff
@@ -0,0 +1,388 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+def dist(a, b, limit=None):
+    return BerghelRoach(a).getDistance(b, limit or len(a) + len(b))
+
+# This is a modification of the original Berghel-Roach edit
```
Comment (fejta) [inline triage/berghelroach.py]: Would this be better as a pydoc comment?

File: triage/index.html [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+Filter (these are regexes):
+<table>
+<label><tr><td>Failure text<td><input type="text" id="filter-text"></label><br>
+<label><tr><td>Job<td><input type="text" id="filter-job"></label><br>
+<label><tr><td>Test<td><input type="text" id="filter-test"></label><br>
+</table>
+</form>
+<div id="summary"></div>
+<div id="clusters">
+  <h2>Loading... <span id="loading-progress"></span></h2>
```
Comment (fejta) [inline triage/index.html]: @rmmh @michelle192837 this is a really nice pattern. Can we add this to testgrid some time this quarter?

File: triage/interactive.js [hunk 21]
```diff
@@ -0,0 +1,255 @@
...
+  req.open('GET', url);
+  req.onload = function(resp) {
+    callback(req);
+  };
+  req.onprogress = onprogress;
+  req.send();
+}
+
+// One-time initialization of the whole page.
+function load() {
```
Comment (fejta) [inline triage/interactive.js]: @michelle192837 this implements the loading behavior for go.k8s.io/triage

File: triage/summarize.py [hunk 22]
```diff
@@ -0,0 +1,369 @@
...
+        {failure_text: [failure_in_cluster_1, failure_in_cluster_2, ...]}
+    '''
+    clusters = {}
+
+    for test in tests:
+        ftext = test['failure_text']
+        fnorm = normalize(ftext)
+        if fnorm in clusters:
+            clusters[fnorm].append(test)
+        else:
```
Comment (fejta) [inline triage/summarize.py]: Thoughts on this one?

Comment (rmmh) [inline triage/summarize.py]: I think if/else is clearer to read than a continue-- less non-local context required. The indentation doesn't isn't very deep here, regardless.

File: triage/interactive.js (hunk 21 above)
Comment (rmmh) [inline triage/interactive.js]: Note that testgrid spends most of its time processing on the server, so it's not possible to integrate loading progress like this.

File: triage/berghelroach.py (hunk 19 above)
Comment (rmmh) [inline triage/berghelroach.py]: Possibly-- but this whole file was translated from Java with a few regexes and some hand editing. That's also why there's a bunch of JavaDoc comments left over.

File: triage/interactive.js (hunk 21 above)
Comment (fejta) [inline triage/interactive.js]: IMO it would still be useful to give feedback that "we're waiting for something from the server" even if we don't report percentages

Comment (fejta) [inline triage/interactive.js]: That will help mitigate confusion about why results aren't showing
